mono/metadata/unity-liveness.c

```diff
@@ -299,11 +391,79 @@ static gboolean mono_traverse_object_internal(MonoObject *object, gboolean isStr
 	return added_objects;
 }
 
+static gboolean mono_validate_object_internal(MonoObject *object, gboolean isStruct, MonoClass *klass, LivenessState *state)
+{
+	int i;
+	MonoClassField* field;
+	MonoClass* p;
+	gboolean added_objects = FALSE;
+
+	g_assert(object);
+
+	// subtract the added offset for the vtable. This is added to the offset even though it is a struct
+	if (isStruct)
+		object--;
+
+	for (p = klass; p != NULL; p = p->parent) {
+		if (p->size_inited == 0)
+			continue;
+		for (i = 0; i < mono_class_get_field_count(p); i++) {
+			field = &p->fields[i];
+			if (field->type->attrs & FIELD_ATTRIBUTE_STATIC)
+				continue;
+
+			if (!mono_field_can_contain_references(field))
+				continue;
+
+			if (MONO_TYPE_ISSTRUCT(field->type)) {
+				char* offseted = (char*)object;
+				offseted += field->offset;
+				if (field->type->type == MONO_TYPE_GENERICINST)
+				{
+					g_assert (field->type->data.generic_class->cached_class);
+					added_objects |= mono_validate_object_internal((MonoObject*)offseted, TRUE, field->type->data.generic_class->cached_class, state);
+				}
+				else
+					added_objects |= mono_validate_object_internal((MonoObject*)offseted, TRUE, field->type->data.klass, state);
+				continue;
+			}
+
+			if (field->offset == -1) {
+				g_assert_not_reached();
+			}
+			else {
+				MonoObject* val = NULL;
+				MonoVTable* vtable = NULL;
+				mono_field_get_value_internal(object, field, &val);
```

Comment:
i think this can be a call to `validate_object_value`?

Reply:
`mono_add_and_validate_object` is doing roughly the same thing unless there's objects getting missed here?